src/compiler/checker.ts

```diff
@@ -20255,25 +20307,142 @@ namespace ts {
             return isCallOrNewExpression(node.parent) && node.parent.expression === node;
         }
 
-        function checkPropertyAccessExpressionOrQualifiedName(node: PropertyAccessExpression | QualifiedName, left: Expression | QualifiedName, right: Identifier) {
+        function getPropertyForPrivateIdentifier(leftType: Type, right: PrivateIdentifier, errorNode: Node | undefined): Symbol | undefined {
+            const privateNameDescription = right.escapedText;
+            const diagName = diagnosticName(right);
+
+            if (!(leftType.symbol.flags & SymbolFlags.Class)) {

```

Comment:
I think a more reliable mechanism would be to:
1. Recursively walk up each containing class (via `getContainingClass`) until you find a class that declares a private name with `privateNameDescription`.
2. If `leftType` does not have a property with that name, then
  1. Get the properties of `leftType` and see if it has *any* private named member with the same description, if one is found then use that information to report an error about the wrong private name.

Reply:
There are other valid types that may not have a `Class` symbol, such as a union two generic instantiations of the containing class, or an interface that extends the containing class. This condition will severely restrict where this can be used.

Reply:
I think Joey incorporated this feedback, would you be up for taking another look?